File: mcp/query_tenant/src/sanitize.ts
import { maskSecret } from './auth.js';

const SENSITIVE_KEYS = new Set([
  'authorization', 'cookie', 'set-cookie', 'x-access-key',
  'x-signature', 'x-timestamp', 'session_token', 'secret_key',
  'access_key', 'token', 'api_key', 'apikey', 'password',
]);

export function sanitizeResponse(data: unknown): unknown {
  if (data == null) return data;
  if (typeof data !== 'object') return data;

  if (Array.isArray(data)) {
    return data.map(sanitizeResponse);
  }

  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
    if (SENSITIVE_KEYS.has(key.toLowerCase())) {
      result[key] = typeof value === 'string' ? maskSecret(value) : '***';
    } else if (typeof value === 'object' && value !== null) {
      result[key] = sanitizeResponse(value);
    } else {
      result[key] = value;
    }
  }
  return result;
}

File: mcp/query_tenant_go/sanitize.go
package main

import "strings"

var sensitiveKeys = map[string]bool{
	"authorization": true, "cookie": true, "set-cookie": true,
	"x-access-key": true, "x-signature": true, "x-timestamp": true,
	"session_token": true, "secret_key": true, "access_key": true,
	"token": true, "api_key": true, "apikey": true, "password": true,
}

func SanitizeResponse(data any) any {
	if data == nil {
		return nil
	}

	switch v := data.(type) {
	case map[string]any:
		result := make(map[string]any, len(v))
		for key, val := range v {
			if sensitiveKeys[strings.ToLower(key)] {
				if s, ok := val.(string); ok {
					result[key] = MaskSecret(s)
				} else {
					result[key] = "***"
				}
			} else {
				result[key] = SanitizeResponse(val)
			}
		}
		return result

	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = SanitizeResponse(item)
		}
		return result

	default:
		return data
	}
}
